### ISSUE TITLE
Fallback source of entropy

### ISSUE THREAD

**Issue by dhardy** [dhardy/rand#22]
This issue is a tracker for the topic *providing an alternative entropy source in case `OsRng` fails*.

Motivation: https://github.com/rust-lang-nursery/rand/issues/180

Existing work: https://github.com/dhardy/rand/pull/21, https://github.com/rust-lang-nursery/rand/pull/181

---

- [`EntropyRng`](https://github.com/rust-lang-nursery/rand/pull/235) provides part of the switching mechanism
- `JitterRng` provides one fallback
- we may still allow applications to provide another entropy source, especially useful for embedded

**Comment by dhardy** [comment]
It would be nice to see:

- [x] An implementation of [the jitterentropy library](https://github.com/smuellerDD/jitterentropy-library) in Rust as a candidate for a *good fallback entropy source*
- [x] Further investigation of the *requirements* for clock-based generators
- [ ] Good entropy estimates for various generators

**Comment by burdges** [comment]
I'd think this should stay in an external crate.  At least folks should not need to roll their own `OsRng` to avoid this.

**Comment by dhardy** [comment]
To avoid it? *Obviously* `OsRng` should *only* return entropy from the OS; one shouldn't have potentially weak entropy used in cryptography without explicitly opting in to it.

Should it be external? Possibly. Initially I thought it would just be a tiny bit of code; now, I'm not so sure.

**Comment by burdges** [comment]
Right, I'm not worried if we're not talking about modifying `OsRng` itself, thanks.  I'd think the linker strip any mutable statics used or whatever if not used, so no worries about where you put it.  :)

**Comment by dhardy** [comment]
#28 implements `JitterRng`.

This RNG has a bunch of diagnostics to check that it can actually get *good* entropy from the clock. In part this is necessary because the code can loop forever if time deltas are constant, but it brings up another question:

*   Do we want a low-quality and a high-quality clock-based entropy source?
*   Do we want very weak entropy vs having nothing?

Possibly the best option is to have `JitterRng` as a high-quality entropy source (possibly useful on some systems without other entropy source) plus another generator as a weak entropy source for `thread_rng` (as a second fall-back, should both `OsRng` and `JitterRng` fail).

**Comment by pitdicker** [comment]
Another option is to make the number of rounds `JitterRng` uses configurable. Now it does 64 rounds of entropy collection for 64 bits. This is very conservative, something like 40 rounds covers desktop processors and servers. Just let it perform something like 8 rounds if a low-quality source is ok.

I also plan to make the timer test run only once, and store the result with `lazy_static`. Now initializing it takes about as much time as generating 8~10 `u64`s, and then 1 `u64`.

What do you propose to use as a third fallback source? I would say that if the timer is not good enough for running `JitterRng`, is is certainly not good enough for a simpler scheme. `RDRAND` is an option though.

**Comment by dhardy** [comment]
Since #54 my branch uses `JitterRng` as a fallback, so I guess this is solved (unless there's a reason this needs to change).

**Comment by burdges** [comment]
Just fyi https://aws.amazon.com/blogs/opensource/better-random-number-generation-for-openssl-libc-and-linux-mainline/

**Comment by dhardy** [comment]
The design of the s2n generator talked about above ↑ sounds really good, in particular the three levels of protection against leaking private random numbers (3 may be overkill, but we don't currently have any of these). We should look into this, whether all via `thread_rng` or using other functions.

Related, we may want to allow users to inject their own source of entropy too (for WASM and `no_std` targets where we currently don't have any source of entropy). ([We mentioned this here](https://github.com/rust-lang-nursery/rand/pull/109#issuecomment-355643860).)

Also see #59.

**Comment by dhardy** [comment]
Much of this was implemented; I opened a separate issue regarding s2n strategies
